src/QsCompiler/DataStructures/SyntaxTokens.fs

```diff
@@ -47,6 +47,7 @@ type QsSymbolKind<'Symbol> =
     /// Used for the arguments of the original method omitted upon functor gen declaration.
     | OmittedSymbols
     /// Used to allow destructs of the form let (_,a) = ...
+    /// Also used for when an alias is omitted in open directives.
```

Comment:
It feels wrong to use `MissingSymbol` in this way, since it otherwise corresponds to a physical `_` in the file. Would it be possible to keep using `QsNullable<QsSymbol>` in the open directive, but just convert the absence of a symbol to empty string later down the line?

Reply:
Dang, I don't like the QsNullable but I'll do whatever is necessary. I will partially revert the second commit

Reply:
I was bringing this up from a design standpoint, by the way, rather than considering whether it's a breaking change for IQ#. I think every case of QsSymbol corresponds to something physically in the file:

- Symbol -> `foo`
- QualifiedSymbol -> `foo.bar`
- SymbolTuple -> `(foo, bar)`
- OmittedSymbols -> `...`
- MissingSymbol -> `_`
- InvalidSymbol -> something unparseable

But for `open foo as bar;` vs `open foo;`, there's nothing to point to, not even a placeholder like `_`. So in that sense it really is an optional symbol (I don't like `QsNullable` much either, but the same argument works for `QsSymbol option`). Does that make sense, or do you disagree?

Reply:
> (I don't like QsNullable much either, but the same argument works for QsSymbol option).

Yeah, that was my problem with it, that we have this rich type system in F# yet we just use a null pointer when you specify an alias of `""`.

In any case, I pushed up a commit that restores the QsNullable. I don't want to take up a lot of y'all's time dying on unnecessary hills

Reply:
QsNullable isn't a null pointer - it has the same type safety as `option`. It's a sum of two cases, `Null` and `Value of 'T`, which is isomorphic to `None` and `Some of 'T` from `option`. The naming is confusing though because `null` (lowercase) really is a null pointer, which we were using in a few places for strings in F#, and I agree that we shouldn't be doing that.

Reply:
I'm not saying it's equivalent to a null pointer in terms of type safety, sorry. I agree that `null` is different from `Null`. I mean that instead of having some conception in an `OpenDirective` fragment of opening a namespace with alias `""` (which is what writing `open X;` should hopefully mean someday), we instead represent that case as `("X", Null)` and downstream we have IsValue/IsNull checks in different places manually rewriting that to `("X", "")`. So yes, you have type safety, but the end result is the same as a null pointer for the alias in that you're checking for a state called "null" every time you want the alias name (three examples: [1][1], [2][2], [3][3]). Those checks don't need to exist if you bake that logic into the QsSymbol type like I originally did.

[1]: https://github.com/microsoft/qsharp-compiler/blob/4c176e5f07e9ae75c368b9a0fb4c632a17e65885/src/QsCompiler/SyntaxProcessor/SymbolOccurrence.fs#L140
[2]: https://github.com/microsoft/qsharp-compiler/blob/4c176e5f07e9ae75c368b9a0fb4c632a17e65885/src/QsCompiler/CompilationManager/EditorSupport/CodeActions.cs#L196
[3]: https://github.com/microsoft/qsharp-compiler/blob/4c176e5f07e9ae75c368b9a0fb4c632a17e65885/src/QsCompiler/CompilationManager/TypeChecking.cs#L136

When I was originally writing this code change, I spent a while trying to trace back where the conception of a null alias was coming from, and found it went all the way back to the fragment kind definition, which is why I changed the definition of an OpenDirective fragment kind. But I understand your point that a fragment is supposed to represent the literal source of the program, and if `""` isn't in the literal source, it shouldn't be a QsSymbol by convention. I'm guessing that despite the name, QsNullable exists as an alternative to having to deal with F# Options in C# and not as a way to represent something called "null"